src/rewriter/rewrite_proof_rule.h

```diff
@@ -0,0 +1,150 @@
+   * @param userFvs The (user-provided) free variable list of the rule. This
+   * is used only to track the original names of the arguments to the rule.
+   * @param fvs The internal free variable list of the rule. Notice these
+   * variables are normalized such that *all* proof rules use the same
+   * variables, per type. In detail, the n^th argument left-to-right of a given
+   * type T is the same for all rules. This is to facilitate parallel matching.
+   * @param cond The conditions of the rule, normalized to fvs.
+   * @param conc The conclusion of the rule, which is an equality of the form
+   * (= t s), where t is specified as rewriting to s. This equality is
+   * normalized to fvs.
+   * @param isFixedPoint Whether the rule should be applied to fixed point in
+   * the strategy
+   * @param isFlatForm Whether the rule is the flat form of the actual rule
+   * with the given id.
+   */
+  void init(DslPfRule id,
+            const std::vector<Node>& userFvs,
+            const std::vector<Node>& fvs,
+            const std::vector<Node>& cond,
+            Node conc,
+            Node context,
+            bool isFlatForm);
+  /** get id */
+  DslPfRule getId() const;
+  /** get name */
+  const char* getName() const;
+  /** Get user variable list */
+  const std::vector<Node>& getUserVarList() const;
+  /** Get variable list */
+  const std::vector<Node>& getVarList() const;
+  /** The context that the rule is applied in */
+  Node getContext() const { return d_context; }
+  /** Does this rule have conditions? */
+  bool hasConditions() const;
+  /** Get (declared) conditions */
+  const std::vector<Node>& getConditions() const;
+  /**
+   * Get the conditions in context { vs -> ss }. This may involve running the
+   * side conditions of this method.
+   */
+  bool getObligations(const std::vector<Node>& vs,
+                      const std::vector<Node>& ss,
+                      std::vector<Node>& vcs) const;
+  /**
+   * Check match, return true if h matches the head of this rule; notifies
+   * the match notify object ntm.
+   *
+   * Note this method is not run as the main matching algorithm for rewrite
+   * proof reconstruction, which considers all rules in parallel.
+   */
+  void getMatches(Node h, expr::NotifyMatch* ntm) const;
+  /** Get conclusion of the rule */
+  Node getConclusion() const;
+  /** Get conclusion of the rule for the substituted terms ss */
+  Node getConclusionFor(const std::vector<Node>& ss) const;
+
+  /**
+   * Is variable explicit? An explicit variable is one that does not occur
+   * in a condition and thus its value must be specified in a proof.
+   */
+  bool isExplicitVar(Node v) const;
+  /**
+   * Get list context. This returns the parent kind of the list variable v.
+   * For example, for
+   *   (define-rule bool-or-true ((xs Bool :list) (ys Bool :list))
+   *      (or xs true ys) true)
+   * The variable xs has list context `OR`.
+   */
+  Kind getListContext(Node v) const;
```

Comment:
clarify the behavior when v is in an ambiguous context

Reply:
Thanks for clarifying. But in that case we should never get to the `return UNDEFINED_KIND;` part of the implementation, right? You could put an assertion there then (or an `Unreachable`).

Reply:
I believe this is the case where there is no list context, e.g. it is not a list variable. I'll clarify.